### ISSUE TITLE
incorrect header check when zlib decompression in http

### ISSUE THREAD

**Issue by AdamMagaluk** [apigee/trireme#181]
In some cases `incorrect header check` is thrown when decompressing gzip http(s) data.

We noticed this when using `superagent` node_module as it sends `Accept-Encoding: gzip,deflate` by default.

Working on getting a script to reproduce this outside of the production envioronment. But using the production endpoint and the `apigee-edge-like-launcher-0.9.0.jar`.

More details to come.

**Comment by AdamMagaluk** [comment]
So it appears that in some cases the java Trireme Buffer can have a offset that points to the actual position in the underlying memory. This is reflected in Buffer as the property `bufOffset`.  Getting an index value looks like this:

[io/apigee/trireme/core/modules/Buffer.java](https://github.com/apigee/trireme/blob/master/core/src/main/java/io/apigee/trireme/core/modules/Buffer.java#L782)
```java
public int get(int index) {
  return (int)buf[index + bufOffset] & 0xff;
}
```

In ZLib.java Trireme converts the `Buffer.BufferImpl` to a ByteBuffer using `chunk.getBuffer()` which calls `return ByteBuffer.wrap(buf, bufOffset, bufLength);` In that process the `offset` property on `ByteBuffer` which `HeapByteBuffer.java` uses with:

```java
protected int ix(int i) {
  return i + offset;
}

public byte get() {
  return hb[ix(nextGetIndex())];
}

public byte get(int i) {
  return hb[ix(checkIndex(i))];
}
```

Zlib uses the new `ByteBuffer` to check if the first two bytes equal the magic header gzip uses. [see](https://github.com/apigee/trireme/blob/master/kernel/src/main/java/io/apigee/trireme/kernel/util/GZipHeader.java#L177-L187)

Because the offset is not set on the `ByteBuffer` it fails to see the header and errors out.

To reproduce:

trireme test.js
```node
var zlib = require('zlib');
var fs = require('fs');
var Duplex = require('stream').Duplex;

var buf = new Buffer([0x1f, 0x8b, 0x08, 0x00, 0x9a, 0x2d, 0xb0, 0x5a, 0x00, 0x03, 0x4b, 0x4c, 0x4a, 0xe6, 0x02, 0x00, 0x4e, 0x81, 0x88, 0x47, 0x04, 0x00, 0x00, 0x00]);

var unzip = zlib.createUnzip();
var stream = new Duplex();
var stream2 = new Duplex();
stream.push(buf);
stream.push(null);

stream.pipe(stream2).pipe(unzip).pipe(process.stdout);
```

**Comment by AdamMagaluk** [comment]
The simplest way to illustrate the issue is using the following snippet. This is essentially how Trireme's `modules/Buffer` works storing the underlying data in `byte[]` with a offset/len index since the underlying array may have extra data.

When Triremes `public ByteBuffer getBuffer()` is called it calls ByteBuffer.wrap with the offset but that does not work properly.

```java
    @Test
    public void testBufferBehaviour()
    {
        int len = 10;
        byte[] buf = new byte[len];
        for (int i=0;i<len; i++) {
            buf[i] = (byte) i;
        }
        int offset = 4;

        ByteBuffer byteBuffer = ByteBuffer.wrap(buf, offset, buf.length-offset);
        byte first = byteBuffer.get(0);
        byte second = byteBuffer.get(1);
        assertEquals(0x4, first);
        assertEquals(0x5, second);
    }
```
I'm not sure Trireme uses the same concept of offset as ByteBuffer does or ByteBuffer is broken (unlikely). If you look at the constructor to `HeapByteBuffer` which wrap calls it calls:

```java
super(-1, off, off + len, buf.length, buf, 0);
```

But that super is and `offset` is always set to 0.

```java
// Creates a new buffer with the given mark, position, limit, capacity,
    // backing array, and array offset
    //
    ByteBuffer(int mark, int pos, int lim, int cap,   // package-private
                 byte[] hb, int offset)
    {
        super(mark, pos, lim, cap);
        this.hb = hb;
        this.offset = offset;
    }
```

**Comment by AdamMagaluk** [comment]
Per ByteBuffer's documentation we are using the `offset` in wrap wrong. That refers to the position the ByteBuffer is at but it specifices the byte array offset will always be zero.

```
```

**Comment by AdamMagaluk** [comment]
Fixed in #182
